cloudinit/net/dhcp.py

```diff
@@ -503,62 +551,275 @@ def get_latest_lease(lease_dir, lease_file_regex):
-                        latest_address = dhcptok
-        return latest_address
-
-
-class Dhcpcd:
+        lease_file = self.get_newest_lease_file_from_distro(distro)
+        if lease_file:
+            content: str
+            content = util.load_file(lease_file)  # pyright: ignore
+            if content:
+                for lease in reversed(self.parse_leases(content)):
+                    server = lease.get(key)
+                    if server:
+                        return server
+
+
+class Dhcpcd(DhcpClient):
     client_name = "dhcpcd"
 
-    def __init__(self):
-        raise NoDHCPLeaseMissingDhclientError("Dhcpcd not yet implemented")
+    def dhcp_discovery(
+        self,
+        interface: str,
+        dhcp_log_func: Optional[Callable] = None,
+        distro=None,
+    ) -> Dict[str, Any]:
+        """Run dhcpcd on the interface without scripts/filesystem artifacts.
+
+        @param interface: Name of the network interface on which to send a
+            dhcp request
+        @param dhcp_log_func: A callable accepting the client output and
+            error streams.
+        @param distro: a distro object for network interface manipulation
+        @return: dict of lease options representing the most recent dhcp lease
+            parsed from the dhclient.lease file
+        """
+        LOG.debug("Performing a dhcp discovery on %s", interface)
+        sleep_time = 0.01
+        sleep_cycles = int(self.timeout / sleep_time)
+
+        # dhcpcd needs the interface up to send initial discovery packets
+        # Generally dhclient relies on dhclient-script PREINIT action to bring
+        # the link up before attempting discovery. Since we are using
+        # -sf /bin/true, we need to do that "link up" ourselves first.
+        distro.net_ops.link_up(interface)
+        try:
+            command = [
+                self.dhcp_client_path,  # pyright: ignore
+                "--ipv4only",  # only attempt configuring ipv4
+                "--waitip",  # wait for ipv4 to be configured
+                "--persistent",  # don't deconfigure when dhcpcd exits
+                "--noarp",  # don't be slow
+                "--script=/bin/true",  # disable hooks
+                interface,
+            ]
+            out, err = subp.subp(
+                command,
+                timeout=self.timeout,
+            )
+            if dhcp_log_func is not None:
+                dhcp_log_func(out, err)
+            lease = self.get_newest_lease(interface)
+            # Attempt cleanup and leave breadcrumbs if it fails, but return
+            # the lease regardless of failure to clean up dhcpcd.
+            if lease:
+                # pid file arrival can take very long, and stdout might contain
+                # the PID, so attempt to use a stdout parsed PID
+                for line in reversed(out.split("\n")):
+                    if "forked to background, child pid " in line:
```

Comment:
> What is the downside of using a more foolproof method to determine the dhcpcd PID by reading the correct /run/dhcpcd/*pid file? 

I revived this code because you reported that it was taking >10s for cloud-init to get its lease - this was an attempt to make timeout faster as a stopgap. Given that you've identified the root cause I agree, we can toss this logic.